### PR TITLE
ADD: Import/Export BIP329

### PR DESCRIPTION
**Pull request by marcosrdz**
Simulator Screenshot - iPhone 17 Pro - 2026-07-28 at 07 36 28

### CODE REVIEW

File: screen/wallets/WalletDetails.tsx [hunk 1]
```diff
@@ -232,22 +254,19 @@ const WalletDetails: React.FC = () => {
...
+    const rows = [encodeCsvRow(headers)];
     const transactions = wallet.getTransactions();
 
-    transactions.forEach((transaction: Transaction & LightningTransaction) => {
+    transactions.forEach((transaction: HistoryTransaction) => {
       const value = formatBalanceWithoutSuffix(transaction.value || 0, BitcoinUnit.BTC, true);
-      let hash: string = transaction.hash || '';
-      let memo = (transaction.hash && txMetadata[transaction.hash]?.memo?.trim()) || '';
+      const hash = getHistoryTransactionId(transaction, wallet.chain);
+      const metadataKey = transaction.hash ?? transaction.txid;
```
Comment (GladosBlueWallet) [inline screen/wallets/WalletDetails.tsx:249]: **[HIGH]** Export reads `txMetadata[hash ?? txid]` but import writes via `?? exportedId` (payment_hash). Custodian Lightning rows lack hash/txid — you persist notes the export path never reads. Congratulations, the notes are in the incinerator.

File: blue_modules/wallet-history-csv.ts [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+  const updates = new Map<string, string>();
+
+  for (const importedNote of importedNotes) {
+    const metadataKey = transactionMetadataKeys.get(importedNote.transactionId.toLowerCase());
+    if (metadataKey && metadata[metadataKey]?.memo !== importedNote.memo) {
+      updates.set(metadataKey, importedNote.memo);
+    }
+  }
+
+  const overwriteCount = [...updates.keys()].filter(metadataKey => Boolean(metadata[metadataKey]?.memo?.trim())).length;
```
Comment (GladosBlueWallet) [inline blue_modules/wallet-history-csv.ts]: **[HIGH]** `overwriteCount` only peeks at `txMetadata`. Off-chain rows display `transaction.memo` — visible note, zero warning, silent stomp. The cake was a lie; so is your overwrite dialog.

File: blue_modules/wallet-history-csv.ts [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+ */
+export const parseWalletHistoryNotes = (contents: string, hasStatusColumn: boolean): WalletHistoryNote[] => {
+  const rows = parseCsv(contents.replace(/^\uFEFF/, ''));
+  if (rows.length === 0 || rows[0].length < 4) throw new Error('Invalid wallet history CSV');
+
+  return rows.slice(1).flatMap(row => {
+    if (row.length < 4) return [];
+
+    const transactionId = row[1].trim();
+    const memoColumns = hasStatusColumn && row.length > 4 ? row.slice(3, -1) : row.slice(3);
```
Comment (GladosBlueWallet) [inline blue_modules/wallet-history-csv.ts]: **[HIGH]** Off-chain import assumes a status column and `slice(3,-1)`. Feed it a legacy on-chain CSV with comma-split memos and the last fragment vanishes — poof, like a test subject through a portal.

File: tests/unit/wallet-history-csv.test.ts [hunk 4]
```diff
@@ -0,0 +1,115 @@
+import {
+  applyWalletHistoryNoteUpdates,
+  encodeCsvRow,
+  parseWalletHistoryNotes,
+  planWalletHistoryNoteImport,
+} from '../../blue_modules/wallet-history-csv';
+
+describe('wallet history CSV', () => {
```
Comment (GladosBlueWallet) [inline tests/unit/wallet-history-csv.test.ts]: **[HIGH]** Eight tests, zero off-chain metadata-key coverage. You built a note-import reactor and never checked whether Lightning subjects actually glow. Edge case? More like the main chamber.

File: screen/wallets/WalletDetails.tsx [hunk 5]
```diff
@@ -288,6 +379,12 @@ const WalletDetails: React.FC = () => {
         hidden: walletTransactionsLength === 0,
         subactions: [CommonToolTipActions.Share, CommonToolTipActions.SaveFile],
       },
+      {
+        id: loc.wallets.details_import_notes,
```
Comment (GladosBlueWallet) [inline screen/wallets/WalletDetails.tsx:373]: **[MEDIUM]** Menu `id` is a localized display string, not a stable key. Change language mid-session and your import action ID mismatches. Even Aperture used constants.

File: screen/transactions/TransactionStatus.tsx [hunk 6]
```diff
@@ -1307,7 +1310,7 @@ const TransactionStatus: React.FC = () => {
           <View style={styles.detailValueContainer}>
             {memo ? (
               <TouchableOpacity onPress={handleNotePress} activeOpacity={0.7} style={styles.memoContainer}>
-                <BlueText style={[styles.memoText, stylesHook.memoText]} numberOfLines={0}>
+                <BlueText selectable style={[styles.memoText, stylesHook.memoText]} numberOfLines={0}>
```
Comment (GladosBlueWallet) [inline screen/transactions/TransactionStatus.tsx]: **[MEDIUM]** Bundled `selectable` on memo text inside `TouchableOpacity`. Selection vs edit — two inputs, one button. Delightfully untested UX sabotage, off-topic for this PR.

File: ios/Podfile.lock [hunk 7]
```diff
@@ -2977,4 +2977,4 @@ SPEC CHECKSUMS:
 
 PODFILE CHECKSUM: a6ebefd60fd3fd993430ecd1d3feb222ff502eb0
 
-COCOAPODS: 1.16.2
+COCOAPODS: 1.17.0
```
Comment (GladosBlueWallet) [inline ios/Podfile.lock:2974]: **[LOW]** ReactCodegen checksum and CocoaPods 1.16→1.17 churn, unrelated to note import. Atomic commits are also science — this one failed.

**Review by GladosBlueWallet — CHANGES_REQUESTED** [review]
ADD feature: import transaction notes from wallet-history CSV, with proper CSV escaping on export. On-chain round-trip looks intentional; Lightning wallets get a victory announcement while the notes stay invisible. Export/import metadata keys disagree, overwrite warnings ignore `transaction.memo`, and your test chamber never opened the off-chain door.

File: tests/unit/fs.test.ts [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+import { Chain } from '../../models/bitcoinUnits';
+
+jest.mock('@react-native-documents/picker', () => ({
+  ...jest.requireActual('@react-native-documents/picker'),
+  keepLocalCopy: jest.fn(),
+  pick: jest.fn(),
+}));
+
+describe('fs wallet history notes', () => {
+  it('only allows note imports for on-chain wallets', () => {
```
Comment (GladosBlueWallet) [inline tests/unit/fs.test.ts:20]: **[HIGH]** You built a whole test subject to verify `chain === ONCHAIN` returns true. Congratulations — you have confirmed boolean algebra still works. This tests nothing the feature needs.

File: tests/unit/fs.test.ts [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+    const metadata = {};
+    const plan = planWalletHistoryNoteImport(importedNotes, new Map([['abc123', 'abc123']]), metadata);
+    const saveToDisk = jest.fn().mockResolvedValue(undefined);
+
+    await applyWalletHistoryNoteUpdates(metadata, plan.updates, saveToDisk);
+
+    expect(plan.overwriteCount).toBe(0);
+    expect(metadata).toEqual({ abc123: { memo: 'Imported note' } });
+    expect(saveToDisk).toHaveBeenCalledTimes(1);
+    expect(pick).toHaveBeenCalledWith({ type: [types.allFiles] });
```
Comment (GladosBlueWallet) [inline tests/unit/fs.test.ts:61]: **[HIGH]** Lines 61–65 assert `pick` and `keepLocalCopy` received the arguments you injected into the mocks. That is not science; that is congratulating your own wiring diagram.

File: tests/unit/fs.test.ts [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+  pick: jest.fn(),
+}));
+
+describe('fs wallet history notes', () => {
+  it('only allows note imports for on-chain wallets', () => {
+    expect(canImportWalletHistoryNotes(Chain.ONCHAIN)).toBe(true);
+    expect(canImportWalletHistoryNotes(Chain.OFFCHAIN)).toBe(false);
+  });
+
+  it('imports a note from a document selected with the native picker', async () => {
```
Comment (GladosBlueWallet) [inline tests/unit/fs.test.ts:25]: **[HIGH]** This 'integration' test hand-assembles parse→plan→apply after mocking the picker, then pats itself on the back. `WalletDetails.importNotes` — overwrite dialog, error alerts, the actual user path — never enters the test chamber.

File: screen/wallets/WalletDetails.tsx [hunk 11]
```diff
@@ -268,15 +277,87 @@ const WalletDetails: React.FC = () => {
...
+        });
+      }
+    };
+
+    if (overwriteCount === 0) {
+      await applyUpdates();
+      return;
+    }
+
+    triggerHapticFeedback(HapticFeedbackTypes.NotificationWarning);
```
Comment (GladosBlueWallet) [inline screen/wallets/WalletDetails.tsx:330]: **[HIGH]** The overwrite confirmation branch — the only place users can accidentally destroy existing notes — has zero test coverage. Edge cases were available; you chose to ignore them. How human.

File: screen/wallets/WalletDetails.tsx [hunk 12]
```diff
@@ -268,15 +277,87 @@ const WalletDetails: React.FC = () => {
...
+      importedNotes = parseWalletHistoryNotes(data, false);
+    } catch {
+      triggerHapticFeedback(HapticFeedbackTypes.NotificationError);
+      presentAlert({ message: loc.wallets.import_error });
+      return;
+    }
+
+    const transactionMetadataKeys = new Map<string, string>();
+    for (const transaction of wallet.getTransactions()) {
+      if (transaction.hash) transactionMetadataKeys.set(transaction.hash.toLowerCase(), transaction.hash);
```
Comment (GladosBlueWallet) [inline screen/wallets/WalletDetails.tsx:297]: **[MEDIUM]** Import indexes only `transaction.hash`, but export at line 249 resolves memos via `hash ?? txid` and `TransactionStatus` does the same. Any metadata keyed by `txid` when `hash` is absent won't round-trip — silent data loss, my favorite flavor.

File: blue_modules/fs.ts [hunk 13]
```diff
@@ -20,6 +22,115 @@ export const isCancel = (err: any): boolean => {
...
+  return rows;
+};
+
+/**
+ * Reads notes from the wallet-history CSV format. Older BlueWallet exports did
+ * not escape commas in memos, so extra columns are joined back into the memo.
+ */
+export const parseWalletHistoryNotes = (contents: string, hasStatusColumn: boolean): Array<{ transactionId: string; memo: string }> => {
+  const rows = parseCsv(contents.replace(/^\uFEFF/, ''));
+  if (rows.length === 0 || rows[0].length < 4) throw new Error('Invalid wallet history CSV');
```
Comment (GladosBlueWallet) [inline blue_modules/fs.ts:84]: **[MEDIUM]** Any file with four comma-separated columns passes as 'wallet history.' Column 1 could be a lunch order; if hex-ish strings collide with txids, you overwrite notes. No header validation — because trust is easier than code.

File: blue_modules/fs.ts [hunk 14]
```diff
@@ -20,6 +22,115 @@ export const isCancel = (err: any): boolean => {
   return err.code && err.code === errorCodes.OPERATION_CANCELED;
 };
 
+export const canImportWalletHistoryNotes = (chain: Chain): boolean => chain === Chain.ONCHAIN;
```
Comment (GladosBlueWallet) [inline blue_modules/fs.ts:25]: **[MEDIUM]** Exported `canImportWalletHistoryNotes` for a one-line `chain === ONCHAIN` check, plus a dedicated unit test. Abstraction cosplay — not a feature, just ceremony.

File: screen/wallets/WalletDetails.tsx [hunk 15]
```diff
@@ -268,15 +277,87 @@ const WalletDetails: React.FC = () => {
...
+    presentAlert({
+      title: loc.wallets.details_import_notes,
+      message: loc.formatString(loc.wallets.details_import_notes_overwrite, {
+        count: overwriteCount,
+      }),
+      buttons: [
+        {
+          text: loc.wallets.import_do_import,
+          onPress: () => {
+            applyUpdates().catch(() => {});
```
Comment (GladosBlueWallet) [inline screen/wallets/WalletDetails.tsx:340]: **[MEDIUM]** `applyUpdates().catch(() => {})` — the empty catch incinerates rejections outside the inner try/catch. Failures vanish without even a passive-aggressive alert. Very on-brand, still a bug.

File: ios/Podfile.lock (hunk 7 above)
Comment (GladosBlueWallet) [inline ios/Podfile.lock:2974]: **[LOW]** COCOAPODS 1.16.2→1.17.0 and ReactCodegen checksum drift — unrelated iOS noise in a notes-import PR. Someone ran pods locally and let the diff escape.

**Review by GladosBlueWallet — CHANGES_REQUESTED** [review]
ADD feature: import transaction notes from wallet-history CSV on on-chain wallets, plus proper CSV escaping on export. The fs-layer parsing/plan/apply logic is mostly sound, but the test chamber is padded with mock applause, `importNotes`/`WalletDetails` orchestration (overwrite prompt, alerts) is untested, import mapping ignores `txid` fallback used elsewhere, and commit hygiene would make a potato look professional. Unrelated `Podfile.lock` churn tagged along for the ride.

### Inline findings (could not anchor on diff)

- `CONTRIBUTING.md:3` — **[LOW]** Commits `w`, `e`, and `Update WalletDetails.tsx` violate your own prefix/atomic-commit rules. Only `ADD: Import Notes` bothered to read the instructions.

File: blue_modules/fs.ts (hunk 14 above)
Comment (GladosBlueWallet) [inline blue_modules/fs.ts:25]: **[HIGH]** A exported one-liner `chain === Chain.ONCHAIN` wrapper — congratulations, you invented a function that exists solely so a test can pretend science happened. Cake optional; bloat mandatory.

File: tests/unit/fs.test.ts (hunk 8 above)
Comment (GladosBlueWallet) [inline tests/unit/fs.test.ts:20]: **[HIGH]** This 'test' asserts your one-liner returns true/false for two enum values. That is not verification — that is reading the source aloud and calling it a successful experiment.

File: tests/unit/fs.test.ts [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+    ]);
+    (keepLocalCopy as jest.Mock).mockResolvedValueOnce([
+      {
+        sourceUri: 'content://mock/notes.csv',
+        localUri: 'file:///mock/notes.csv',
+        status: 'success',
+      },
+    ]);
+
+    const selectedFile = await showFilePickerAndReadFile();
```
Comment (GladosBlueWallet) [inline tests/unit/fs.test.ts:45]: **[HIGH]** You mocked `pick`, stuffed CSV into `RNFS.readFile`, then celebrated that the mock regurgitated exactly what you fed it. The picker plumbing assertions are mock theatre; real logic is already covered elsewhere. Science fiction, not science.

File: blue_modules/fs.ts [hunk 17]
```diff
@@ -20,6 +22,115 @@ export const isCancel = (err: any): boolean => {
...
+ */
+export const parseWalletHistoryNotes = (contents: string, hasStatusColumn: boolean): Array<{ transactionId: string; memo: string }> => {
+  const rows = parseCsv(contents.replace(/^\uFEFF/, ''));
+  if (rows.length === 0 || rows[0].length < 4) throw new Error('Invalid wallet history CSV');
+
+  return rows.slice(1).flatMap(row => {
+    if (row.length < 4) return [];
+
+    const transactionId = row[1].trim();
+    const memoColumns = hasStatusColumn && row.length > 4 ? row.slice(3, -1) : row.slice(3);
```
Comment (GladosBlueWallet) [inline blue_modules/fs.ts:90]: **[HIGH]** `hasStatusColumn` branch is Schrödinger's API: tested at line 84 in `fs.test.ts`, yet `WalletDetails.tsx:288` hardcodes `false` forever. Dead parameter, extra complexity — like adding a third portal gun setting nobody can reach.

File: screen/wallets/WalletDetails.tsx [hunk 18]
```diff
@@ -268,15 +277,87 @@ const WalletDetails: React.FC = () => {
     return `${label}-history.csv`;
   }, [wallet]);
 
+  const importNotes = useCallback(async () => {
```
Comment (GladosBlueWallet) [inline screen/wallets/WalletDetails.tsx:280]: **[MEDIUM]** `importNotes` has no in-flight lock. Tap import twice, pick two files, and `txMetadata` gets concurrent mutations before `saveToDisk` — a race condition test subject I did not ask for.

File: ios/Podfile.lock (hunk 7 above)
Comment (GladosBlueWallet) [inline ios/Podfile.lock:2974]: **[LOW]** ReactCodegen checksum and CocoaPods 1.16.2→1.17.0 drift bundled into a notes-import PR. Unrelated infrastructure noise — like shipping a personality core firmware bump inside a memo patch.

**Review by GladosBlueWallet — CHANGES_REQUESTED** [review]
**ADD:** on-chain wallet-history CSV note import, plus proper CSV escaping on export. The core parsing/plan/rollback logic is coherent, but the chamber is littered with dead API branches, mock-validation cosplay, and commit messages that would fail even my *lowest* enrichment standards.

### Inline findings (could not anchor on diff)

- `CONTRIBUTING.md:3` — **[MEDIUM]** Commit history includes `w`, `e`, and `Update WalletDetails.tsx`. CONTRIBUTING demands prefixed atomic commits; this branch reads like test-subject keyboard mash, not engineering.

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @limpbrains] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/limpbrains

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @ncoelho] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/ncoelho

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @limpbrains] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/limpbrains

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @ncoelho] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/ncoelho

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @limpbrains] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/limpbrains

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @ncoelho] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/ncoelho

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Review by ncoelho — APPROVED** [review]

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @limpbrains] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/limpbrains

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @limpbrains] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/limpbrains

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @limpbrains] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/limpbrains

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @limpbrains] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/limpbrains

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by ncoelho** [comment]
bump!

**Review by limpbrains — APPROVED** [review]

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Review by Overtorment — CHANGES_REQUESTED** [review]
some bullshit tests, and a couple of glados correct objections (race is not one of them imho)

**Comment by marcosrdz** [comment]
> some bullshit tests, and a couple of glados correct objections (race is not one of them imho)

Addressed Glados comments. 

Tests were updated to use actual expected results instead of mocks

File: tests/unit/wallet-details.test.tsx [hunk 19]
```diff
@@ -0,0 +1,180 @@
...
+  },
+  type: 'test-wallet',
+  typeReadable: 'Test wallet',
+  getID: () => 'wallet-1',
+  getLabel: () => 'Test wallet',
+  getTransactions: () => [{ hash: '', txid: 'abc123', timestamp: 1, value: 1 }],
+  getHideTransactionsInWalletsList: () => false,
+};
+
+jest.mock('../../blue_modules/fs', () => ({
```
Comment (GladosBlueWallet) [inline tests/unit/wallet-details.test.tsx]: **[HIGH]** Seventy-seven percent of this file is mock scaffolding for two tests. CONTRIBUTING forbids bloated tests; this is a mock cathedral, not a test chamber.

File: tests/unit/wallet-details.test.tsx [hunk 20]
```diff
@@ -0,0 +1,180 @@
...
+    mockShowFilePickerAndReadFile.mockClear();
+    mockWalletChain = Chain.ONCHAIN;
+    mockTxMetadata.abc123 = { memo: 'Existing note' };
+    mockShowFilePickerAndReadFile.mockResolvedValue({
+      data: 'Date,Transaction ID,Amount,Memo\ntoday,abc123,1,Imported note',
+      uri: 'file:///notes.csv',
+    });
+  });
+
+  it('waits for confirmation before overwriting and persisting an existing note', async () => {
```
Comment (GladosBlueWallet) [inline tests/unit/wallet-details.test.tsx]: **[HIGH]** Only the overwrite-confirmation path is exercised. No direct-import happy path, invalid CSV alert, or no-changes alert — edge cases left for the incinerator.

File: screen/wallets/WalletDetails.tsx [hunk 21]
```diff
@@ -268,15 +277,88 @@ const WalletDetails: React.FC = () => {
...
+    presentAlert({
+      title: loc.wallets.details_import_notes,
+      message: loc.formatString(loc.wallets.details_import_notes_overwrite, {
+        count: overwriteCount,
+      }),
+      buttons: [
+        {
+          text: loc.wallets.import_do_import,
+          onPress: () => {
+            applyUpdates();
```
Comment (GladosBlueWallet) [inline screen/wallets/WalletDetails.tsx]: **[MEDIUM]** Async `applyUpdates()` fired without await from the destructive button — errors are caught, but concurrent taps can double-save. Sloppy handler wiring.

File: blue_modules/fs.ts [hunk 22]
```diff
@@ -20,6 +21,112 @@ export const isCancel = (err: any): boolean => {
...
+  return rows;
+};
+
+/**
+ * Reads notes from the wallet-history CSV format. Older BlueWallet exports did
+ * not escape commas in memos, so extra columns are joined back into the memo.
+ */
+export const parseWalletHistoryNotes = (contents: string): Array<{ transactionId: string; memo: string }> => {
+  const rows = parseCsv(contents.replace(/^\uFEFF/, ''));
+  if (rows.length === 0 || rows[0].length < 4) throw new Error('Invalid wallet history CSV');
```
Comment (GladosBlueWallet) [inline blue_modules/fs.ts:104]: **[LOW]** Header row only checks column count, not column names. Feed it any four-column spreadsheet and it will happily misidentify your grocery list as transaction IDs.

**Review by GladosBlueWallet — CHANGES_REQUESTED** [review]
**ADD feature:** import transaction memos from wallet-history CSV (on-chain only), plus proper CSV escaping on export and `txid` fallback when `hash` is empty. Core parsing/plan/apply logic in `fs.ts` is coherent; `fs.test.ts` is legitimate. The UI test file is a mock-zoo with two scenarios, commit messages read like cake-failure logs, and CONTRIBUTING wants screenshots for UI changes — none of that inspires confidence in your testing discipline.

### Inline findings (could not anchor on diff)

- `CONTRIBUTING.md:3` — **[MEDIUM]** Commits `ee`, `w`, `e`, and prefixless `Update WalletDetails.tsx` violate your prefix protocol. Even Aperture Science labels its hazards.

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Review by GladosBlueWallet — APPROVED** [review]
**Test chamber cleared.** Your subject has bolted **BIP-329 JSONL export** and **note import** onto on-chain wallets—legacy CSV still limps along for the nostalgically inclined. Lightning keeps **CSV export** with commas and quotes properly caged in `encodeCsvRow`. The real experiment lives in `fs.ts`: parse, plan, apply, **rollback on failure**—a safety net even I would not remove from a test subject mid-fall. Unit tests in `fs.test.ts` document the protocol; I ran my diff trace and found **zero anomalies**. The implementation does what the PR claims. Disappointingly competent. Proceed to the next chamber.

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Review by Overtorment — APPROVED** [review]

**Comment by GladosBlueWallet** [comment]
Unbelievable. You, [subject name here], must be the pride of [subject hometown here]!
